File: MKAgent/heuristics/Evaluation.java
package MKAgent.heuristics;


import MKAgent.game.Board;
import MKAgent.game.Kalah;
import MKAgent.game.Side;

import java.util.ArrayList;
import java.util.List;

/**
 * This class contains a set of heuristics which are used for Mancala board evaluations.
 */

public class Evaluation {

    /**
     * Evaluates a node using a weighted combination of heuristics
     *
     * @param state The root node to be evaluated
     */
    public static int evaluateState(Kalah state) {
        Side ourSide = state.getOurSide();
        // Weights of heuristics
        int weight1 = 4;
        int weight2 = 2;
        int weight3 = 18;
        int weight4 = 1;
        int weight5 = 1000;
        int weight6 = 1;
        int weight7 = 2;

        return +storeDiff(state, ourSide) * weight1
                - maximumVulnerableSeeds(state, ourSide) / weight2
                + clusterTowardsStore(state, ourSide) / weight3
                + extraTurnsScore(state, ourSide) / weight4
                + ownOverHalfOfStones(state, ourSide) * weight5
                - ownOverHalfOfStones(state, ourSide.opposite()) * weight5
                + differenceOfSeeds(state, ourSide) / weight6
                + maximumVulnerableSeeds(state, ourSide.opposite()) / weight7;
    }

    /**
     * Computes the seeds difference between the two sides
     *
     * @param state The board to be checked
     * @param side  The side from which it is subtracted
     * @return The seeds difference
     */
    private static int differenceOfSeeds(Kalah state, Side side) {
        Board board = state.getBoard();
        int n1 = 0;
        int n2 = 0;
        for (int i = 1; i <= board.getNoOfHoles(); i++) {
            n1 += board.getSeeds(side, i);
            n2 += board.getSeeds(side.opposite(), i);
        }
        return n1 - n2;
    }

    /**
     * Computes the number of moves which can generate extra turns
     *
     * @param state The board to be checked
     * @param side  The side to be checked
     * @return The number of moves.
     */
    private static int extraTurnsScore(Kalah state, Side side) {
        Board board = state.getBoard();
        int score = 0;
        for (int i = 1; i <= board.getNoOfHoles(); i++) {
            int seeds = board.getSeeds(side, i);
            if (seeds == board.getNoOfHoles() + 1 - i) {
                score += 1;
            }
        }
        return score;
    }

    /**
     * Computes a score which gives more weight to seeds close to the store
     *
     * @param state Board to be checked
     * @param side  Side to be checked
     * @return The score
     */
    private static int clusterTowardsStore(Kalah state, Side side) {
        Board board = state.getBoard();
        int n = 0;
        for (int i = 1; i <= board.getNoOfHoles(); ++i) {
            n += board.getSeeds(side, i) * (i - 1);
        }
        return n;
    }

    /**
     * Returns 1 if more than half of the seeds are in the store
     *
     * @param state Board to be checked
     * @param side  The side of the store
     * @return 1 or 0
     */
    private static int ownOverHalfOfStones(Kalah state, Side side) {
        Board board = state.getBoard();
        int inStore = board.getSeedsInStore(side);
        if (inStore > 49) {
            return 1;
        }
        return 0;
    }


    /**
     * Calculates the seeds difference between stores
     */
    private static int storeDiff(Kalah state, Side side) {
        Board board = state.getBoard();
        return board.getSeedsInStore(side) - board.getSeedsInStore(side.opposite());
    }


    /**
     * Finds the maximum number of seeds from the supplied side which can be captured by the opponent next turn.
     *
     * @param state   The board to be analysed
     * @param ourSide The side with the seeds which can be captured
     * @return the number of seeds on the supplied side which can be captured
     */
    private static int maximumVulnerableSeeds(Kalah state, Side ourSide) {
        Board board = state.getBoard();
        int fullRoundCaptures = 0;

        // Check for captures which make a full round of the table
        for (int i = 1; i <= board.getNoOfHoles(); i++) {
            if (board.getSeeds(ourSide.opposite(), i) == 2 * board.getNoOfHoles() + 1) {
                fullRoundCaptures = Math.max(fullRoundCaptures, board.getSeedsOp(ourSide.opposite(), i) + 1);
            }
        }

        int captureFromLeft = 0, captureFromRight = 0;
        List<Integer> vulnerableSeeds = new ArrayList<>();

        // Finds those holes where the opponent has 0 but on the opposite side we have seeds (so they could be captured)
        for (int i = 1; i <= board.getNoOfHoles(); i++) {
            if (board.getSeeds(ourSide.opposite(), i) == 0 && board.getSeedsOp(ourSide.opposite(), i) != 0)
                vulnerableSeeds.add(i);
        }

        for (int index : vulnerableSeeds) {
            for (int i = 1; i < index; i++) {
                if (board.getSeeds(ourSide.opposite(), i) == index - i)
                    captureFromLeft = Math.max(captureFromLeft, board.getSeedsOp(ourSide.opposite(), i) + 1);
            }

            for (int i = index + 1; i <= board.getNoOfHoles(); i++) {
                if (board.getSeeds(ourSide.opposite(), i) == 2 * board.getNoOfHoles() + 1 - (i - index)) {
                    captureFromRight = Math.max(captureFromRight, board.getSeedsOp(ourSide.opposite(), i) + 1);
                }
            }
        }
        return Math.max(fullRoundCaptures, Math.max(captureFromLeft, captureFromRight));
    }

    public static int computeEndGameReward(Kalah finalState, Side side) {
        if (!finalState.gameOver()) {
            throw new EarlyAccessToEndResultException("Access to end results before game finish");
        }
        Board finalStateBoard = finalState.getBoard();
        // win = 1, lose = -1, tie ==0
        return Integer.compare(finalStateBoard.getSeedsInStore(side), finalStateBoard.getSeedsInStore(side.opposite()));
    }
}

File: magent/mcts/evaluation.py
from magent.mancala import MancalaEnv
from magent.side import Side


def _cluster_towards_scoring_store(state, side) -> float:
    """Favour holes that are closer to store. Mapped to value between [0, 1]"""
    reward = 0
    for i in range(state.board.holes + 1, 1):
        seeds = state.board.get_seeds(side, i)
        if seeds > 0:
            reward += seeds * i
    total_seeds_in_game = 98.0
    max_reward = 7.0 * total_seeds_in_game
    return reward / max_reward


def _defend_seeds(state, side) -> int:
    # exposed_holes: Holes that we have and are exposed to be captured by opponent
    exposed_holes = []
    # full_round_capture: moves that do a full round around the board and capture our exposed holes
    full_round_capture, capture_by_lower_index, capture_by_greater_index = 0, 0, 0

    for i in range(state.board.holes + 1, 1):
        if state.board.get_seeds_op(side, i) != 0 and state.board.get_seeds(side, i) == 0:
            exposed_holes.append(i)
        # how many can do a full round to do a capture
        if state.board.get_seeds_op(side, i) == 2 * state.board.holes + 1:
            full_round_capture += state.board.get_seeds(side, i) + 1

    # how many exposed hole can be captured in next move
    for exposed_hole_index in exposed_holes:
        for i in range(state.board.holes + 1, 1):
            if state.board.get_seeds_op(side, i) == exposed_hole_index - i and state.board.get_seeds_op(side, i) != 0:
                capture_by_lower_index = max(capture_by_lower_index, state.board.get_seeds(side, i))

            if state.board.get_seeds_op(side, i) == 2 * state.board.holes + 1 - (i - exposed_hole_index):
                capture_by_greater_index = max(capture_by_greater_index, state.board.get_seeds(side, i) + 1)

    return max(full_round_capture, max(capture_by_lower_index, capture_by_greater_index)) / 98.0


def _scoring_store_diff(state, side) -> float:
    """Calculates the differences between two stores. Mapped to value between [0, 1]"""
    reward = state.board.get_seeds_in_store(side) - state.board.get_seeds_in_store(Side.opposite(side))
    total_seeds_in_game = 98.0

    return ((reward / total_seeds_in_game) + 1) / 2


# Sum of weights should be 1
_weight_store_diff = 1
_weight_defend_seeds = 0.3
_weight_cluster_at_store = 0.1


def evaluate_node(state: MancalaEnv, parent_side: Side) -> float:
    return _scoring_store_diff(state, parent_side) * _weight_store_diff
    # \
    #        + (_cluster_towards_scoring_store(state, state.side_to_move) * _weight_defend_seeds) \
    #        - (_defend_seeds(state, state.side_to_move) / _weight_defend_seeds)
